book/developers/exex/tracking-state.md

```diff
@@ -0,0 +1,193 @@
+   so that if the channel of notifications has no value ready, we will instantly return `Poll::Pending` from our Future.
+1. We initialize and return the `MyExEx` struct directly in the `install_exex` method, because it's a Future.
+
+With all that done, we're now free to add more fields to our `MyExEx` struct, and track some state in them.
+
+### Adding state
+
+Our ExEx will count the number of transactions in each block and log it to the console.
+
+```rust,norun,noplayground,ignore
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+
+use reth::{api::FullNodeComponents, primitives::BlockNumber};
+use reth_exex::{ExExContext, ExExEvent};
+use reth_node_ethereum::EthereumNode;
+use reth_tracing::tracing::info;
+
+struct MyExEx<Node: FullNodeComponents> {
+    ctx: ExExContext<Node>,
+    /// First block that was committed since the start of the ExEx.
+    first_block: Option<BlockNumber>,
+    /// Total number of transactions committed.
+    transactions: u64,
+}
+
+impl<Node: FullNodeComponents> MyExEx<Node> {
+    fn new(ctx: ExExContext<Node>) -> Self {
+        Self {
+            ctx,
+            first_block: None,
+            transactions: 0,
+        }
+    }
+}
+
+impl<Node: FullNodeComponents> Future for MyExEx<Node> {
+    type Output = eyre::Result<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        while let Some(notification) = ready!(this.ctx.notifications.poll_recv(cx)) {
+            if let Some(reverted_chain) = notification.reverted_chain() {
+                this.transactions = this.transactions.saturating_sub(
+                    reverted_chain
+                        .blocks_iter()
+                        .map(|b| b.body.len() as u64)
+                        .sum(),
+                );
+            }
+
+            if let Some(committed_chain) = notification.committed_chain() {
+                this.first_block.get_or_insert(committed_chain.first().number);
+
+                this.transactions += committed_chain
+                    .blocks_iter()
+                    .map(|b| b.body.len() as u64)
+                    .sum::<u64>();
+
+                this.ctx
+                    .events
+                    .send(ExExEvent::FinishedHeight(committed_chain.tip().number))?;
+            }
+
+            if let Some(first_block) = this.first_block {
+                info!(%first_block, transactions = %this.transactions, "Total number of transactions");
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+}
+
+fn main() -> eyre::Result<()> {
+    reth::cli::Cli::parse_args().run(|builder, _| async move {
+        let handle = builder
+            .node(EthereumNode::default())
+            .install_exex("my-exex", |ctx| async move { Ok(MyExEx::new(ctx)) })
+            .launch()
+            .await?;
+
+        handle.wait_for_node_exit().await
+    })
+}
+```
```

Comment:
I don't like that we basically repeat the whole snippet twice. How can we do it better?

Reply:
https://rust-lang.github.io/mdBook/format/mdbook.html#including-files

Reply:
we don't fully repeat the snippet but rather change some parts of it. I meant how we can do better at showing what changed in the code without repeating it all over again, but we can solve it later.